test/ExportVerilog/hw-typedecls.mlir

```diff
@@ -7,4 +7,25 @@ hw.type_scope @__hw_typedecls {
   hw.typedecl @bar : !hw.struct<a: i1, b: i1>
   // CHECK: typedef logic [7:0][0:15] baz;
   hw.typedecl @baz : !hw.uarray<16xi8>
+  // CHECK: typedef logic [31:0] customName;
+  hw.typedecl @qux, "customName" : i32
+  // CHECK: typedef struct packed {foo a; _other_scope_foo b; } nestedRef;
+  hw.typedecl @nestedRef : !hw.struct<a: !hw.typealias<@__hw_typedecls::@foo,i1>, b: !hw.typealias<@_other_scope::@foo,i2>>
```

Comment:
Is there any logic to ensure that the typedefs are defined before they're used?

Reply:
It doesn't have to be part of this PR, but if 'no' it'd be nice to not have to worry about that when/if creating the `type_scope` sections.

Reply:
Currently not, but it should doable to follow up with that support.

Reply:
The output should be valid regardless of where the type_scope is defined in the ir.  This also needs to work correctly with split-verilog output from firtool.

Reply:
I think there are two things here: wherever the type_scope shows up in the IR, we want to emit the typedecls up at the top of the output, before anything uses them. That is how type_scopes emit currently. This also works with split-verilog right now by copying all typedecls to the top of all files. Down the road, we have discussed emitting them to a .svh header that can be included in each file instead.

The second issue is we need to ensure that the typedecls are emitted in an order that makes sure any nested typedecls are defined before referenced. This PR doesn't do anything to ensure that, but it should be possible to do a topological sort of the typedecl dependencies and make sure they are emitted such that any nested typedecls are defined before used.